File: 322-coin-change/322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        minCoins = [amount + 1] * (amount + 1)
        minCoins[0] = 0
        
        for i in range(1, amount + 1):
            for j in coins:
                if i - j >= 0:
                    minCoins[i] = min(minCoins[i], 1 + minCoins[i - j])
        
        return minCoins[-1] if minCoins[-1] != amount + 1 else -1

File: 322-coin-change/322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> minCoins(amount + 1, amount + 1);
        minCoins[0] = 0;
        
        for(int i = 1; i < amount + 1; i++)
            for(auto coin : coins)
                if(i - coin >= 0)
                    minCoins[i] = min(minCoins[i], 1 + minCoins[i - coin]);
        
        return minCoins[amount] == amount + 1 ? -1 : minCoins[amount];
    }
};